clash-cores/test/Test/Cores/Xilinx/DcFifo.hs

```diff
@@ -0,0 +1,146 @@
+  -- | @xs@
+  [a] ->
+  -- | @ys@
+  [a] ->
+  Bool
+orderFifo [] [] = True
+orderFifo (x:xs) (y:ys) | x == y = orderFifo xs ys
+orderFifo xs (_:ys) = orderFifo xs ys
+orderFifo _ _ = False
+
+-- | Generated stalls are large enough that we expect loss, but order should be
+-- preserved.
+prop_fifoOrder :: Property
+prop_fifoOrder = property $ do
+  (xs, wrIn, rdStalls) <- forAll $ genScenario (Gen.int (Range.linear 30 40)) (Gen.int (Range.linear 10 15))
+  H.assert (orderFifo (throughFifo wrIn rdStalls) xs)
+
+-- | Generated stalls are small enough that we don't expect any data loss
+prop_noloss :: Property
+prop_noloss = property $ do
+  (xs, wrIn, rdStalls) <- forAll $ genScenario (Gen.int (Range.linear 7 12)) (Gen.int (Range.linear 7 8))
+  throughFifo wrIn rdStalls === xs
+
+genData :: (KnownNat n, MonadGen m) => m (BitVector n)
+genData = Gen.enumBounded
+
+-- | Mealy machine which stalls reading from the FIFO based on ['Maybe'
+-- 'Int']
+takeState ::
+  (Bool, [Maybe Int]) ->
+  (ResetBusy, Empty, DataCount depth, BitVector n) ->
+  ((Bool, [Maybe Int]), (Maybe (BitVector n), Bool))
+takeState (_, stalls) (True, _, _, _) = ((False, stalls), (Nothing, False))
+takeState (readLastCycle, Just 0:stalls) (_, _, _, d) =
+    ((False, stalls), (nextData, False))
+  where
+    nextData = readLastCycle `orNothing` d
+takeState (readLastCycle, Just n:stalls) (_, _, _, d) =
+    ((False, Just (n-1):stalls), (nextData, False))
+  where
+    nextData = readLastCycle `orNothing` d
+takeState (readLastCycle, stalls) (_, fifoEmpty, _, d) =
+    ((readThisCycle, L.drop 1 stalls), (nextData, readThisCycle))
+  where
+    readThisCycle = not fifoEmpty
+    nextData = readLastCycle `orNothing` d
+
+feedState ::
+  [Either Int (BitVector 32)] ->
+  (ResetBusy, Full, DataCount 4) ->
+  ([Either Int (BitVector 32)], Maybe (BitVector 32))
+feedState xs (True, _, _) = (xs, Nothing)
+feedState [] _ = ([], Nothing)
+feedState (Left 0:xs) (_, _, _) = (xs, Nothing)
+feedState (Left i:xs) (_, _, _) = (Left (i-1):xs, Nothing)
+feedState (Right x:xs) (_, full, _) =
+  if full
+    then (Right x:xs, Nothing)
+    else (xs, Just x)
+
+throughFifo
+  :: [Either Int (BitVector 32)] -- ^ Write data ('Left' 'Int' indicates a stall of duration @i@)
+  -> [Maybe Int] -- ^ Read stalls
+  -> [BitVector 32]
+throughFifo wrDataList rdStalls = rdDataList
+  where
+    clk = clockGen @XilinxSystem
+    rst = resetGen @XilinxSystem
+    ena = enableGen @XilinxSystem
+    wrData =
+      -- The reset to the mealy machine must be the same reset fed to the FIFO
+      mealyB clk rst ena feedState wrDataList (wrRstBusy, wrFull, wrCnt)
+    (rdDataMaybe, rdEna) =
+      mealyB clk rst ena takeState (False, rdStalls) (rdRstBusy, rdEmpty, rdCnt, rdData)
+
+    rdDataList =
+      catMaybes
+        $ sampleN (P.length wrDataList*10)
+        $ bundle rdDataMaybe
+
+    (XilinxFifo wrRstBusy wrFull _ wrCnt rdRstBusy rdEmpty _ rdCnt rdData) =
+      dcFifo defConfig clk clk rst wrData rdEna
```

Comment:
At a first glance it feels really limited to test a _dual clock_ FIFO only with a single clock. No interesting situations where the number of ticks of one clock between the ticks of the other varies (it switches between floor(Ta/Tb) and ceil(Ta/Tb)), no large differences between the two frequencies... I think you should add some more ratios.

I think I understand why you do this in HDL simulation, since you indicated the Haskell model doesn't match the Vivado model (though both should be semantically correct), but that's not an excuse here ;-).

Reply:
Right, I've got some stuff I use on the elastic buffer; I'll try it here and hopefully it'll make things better.

Reply:
I think Peter is right, we should tests with more ratios here. Luckily, it shouldn't be that hard. It should be a matter of doing two things:

 * In `Test.Cores.Xilinx.DcFifo`: Have `throughFifo` take a `Proxy (writeDom :: Domain) -> Proxy (readDom :: Domain)`. Propagate those to the mealy machines. Create a few "interesting" domains with `createDomain` and test some by adding them to `test`.
 * For the HDL simulation: do a very similar thing, but create multiple top test benches. 

Also I'm not sure what an elastic buffer is, if you catch my drift :wink:

Reply:
> For the HDL simulation: do a very similar thing, but create multiple top test benches.

So does the Haskell model match the Vivado model or not? When I initially saw that the HDL only tested a single domain, I figured that was because they do _not_ match and this very benign case did match, so it was the one we could test.

Reply:
It doesn't, but that's on purpose. Xilinx is very explicit in that their RTL model doesn't match the generated hardware, and that you should only rely on the interface - not specific behavior around reset length / cross-domain clocking determinism etc.

We test the Haskell model using Hedgehog, the RTL model using a translatable testbench.

Reply:
Oh, now I'm leaving out the most important part: we figured it wouldn't be worth the effort to match the exact RTL model, as that doesn't match the hardware anyway.

Reply:
That's all good, but that complicates writing an HDL test bench because you're saying the HDL behaviour would be different from the Haskell behaviour, so you have to filter out the differences in your test bench if you want it to succeed in both simulations. And if you filter too much, you're not testing much either.

Unless you're suggesting writing a test bench which purposely fails running it in Haskell (like with `runUntil id testBench`), which I would be opposed to in principle.

Also, unless you're working with a defined phase relationship between the two clocks, you could never have a model that matches reality in a cycle-exact manner with clock crossings, because real clocks do not exhibit ideal behaviour. And, by the way, if you had a defined phase relationship, you could write a more efficient clock crossing function than this FIFO ;-).

I'm aware I'm saying a lot of stuff you already know, but it's good to have it explicitly mentioned here, I think.

Reply:
The existing HDL test bench folds all the checking logic into HDL. So you get a circuit that drives the FIFO, one that samples it, and one that checks all the written values are read. That doesn't depend on very specific timing characteristics, just that the FIFO behaves -handwavely- like `id`.

Reply:
> Unless you're suggesting writing a test bench which purposely fails running it in Haskell (like with runUntil id testBench), which I would be opposed to in principle.

I'm not sure what this means, but the tests all work just fine in Haskell too :).

Reply:
I had typed a response, but I see your edit now :).

Provided that we'll test with multiple clock domains (which is a TODO), we should test all properties you asked for except for data count. For readers at home: both the write and read side get fed an (differently seeded) LSFR to generate stall information. We should get some pretty nice properties out of that.

Testing the data count is a bit annoying, because the "data count" is a bit of a fuzzy concept due to the multiple clock domains. I'm not sure what we could test there.

Reply:
I've just pushed a commit that works in multiple clock domains (Haskell side). 

(there are a couple re-ordering tests where it over/under-flows)

Reply:
Thanks! Same domain test is also interesting, so you can re-add those.

Reply:
Ok just pushed one RTL test which uses multiple clock domains I believe. I will add another soon!

Reply:
Thanks, I think that resolves this conversation.